torch/utils/data/datapipes/map/combinatorics.py

```diff
@@ -64,3 +65,53 @@ def __iter__(self) -> Iterator[T_co]:
 
     def __len__(self) -> int:
         return len(self.datapipe)
+
+@functional_datapipe('drop')
+class DropperMapDataPipe(MapDataPipe[T_co]):
+    r"""
+    Drop columns/elements in input DataPipe via its indices (functional name: ``drop``).
+
+    Args:
+        datapipe: MapDataPipe with columns to be dropped
+        indices: a single column index to be dropped or a list of indices
+
+    Example:
+        >>> from torchdata.datapipes.map import SequenceWrapper, ZipperMapDataPipe
+        >>> dp1 = SequenceWrapper(range(5))
+        >>> dp2 = SequenceWrapper(range(10, 15))
+        >>> dp = dp1.zip(dp2)
+        >>> list(dp)
+        [(0, 10), (1, 11), (2, 12), (3, 13), (4, 14)]
+        >>> drop_dp = dp.drop(1)
+        >>> list(drop_dp)
+        [(0), (1), (2), (3), (4)]
+    """
+    datapipe: MapDataPipe[T_co]
+
+    def __init__(self,
+                 datapipe: MapDataPipe[T_co],
+                 indices: Union[int, List],
+                 ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+        if isinstance(indices, list):
+            self.indices = indices
+        else:
+            self.indices = [indices]
+
+    def __getitem__(self, index) -> T_co:
+        old_item = self.datapipe[index]
+        new_item = deepcopy(old_item)
```

Comment:
For `IterDataPipe`, I am sure that we don't need to `deepcopy`, as each Iterator of `IterDataPipe` should provide [a new copy of data](https://github.com/pytorch/pytorch/blob/b91ff5e361623685799b8ef725a91b756685a9ae/torch/utils/data/datapipes/iter/utils.py#L34-L36) and we enforce to only have single iterator from a `IterDataPipe`. This would prevent in-place modification of the original data.
However, for `MapDataPipe`, I hesitate considering [the copy happens at construction time](https://github.com/pytorch/pytorch/blob/b91ff5e361623685799b8ef725a91b756685a9ae/torch/utils/data/datapipes/map/utils.py#L33-L34) and we don't have a single-iterator pattern to prevent the original data being modified. And, considering when the data is `tuple`, we apparently do a copy by converting it to a `list` and drop item(s) then converting it back to `tuple`. It seems better to me by making the behavior aligned for the cases of `tuple` and `list`. Overall, It makes sense to me to use `deepcopy` here. 
WDYT? cc: @NivekT @VitalyFedyunin 

For the following case
```
dp = SequenceWrapper([[1,2], [3,4], ...])
dp1 = dp.drop(1)
def inplace_fn(d):
    for idx in range(len(d)):
        d[idx] += 1
dp2 = dp.map(inplace_fn)
dp3 = dp1.zip(dp2)
# What would be our expected result?
# Option 1 with deepcopy
# [([1], [2, 3]), ([3], [4, 5]), ...]
# Option 2 no deepcopy (seems weird to me)
# [([1], [2]), ([3], [4]), ...]
```

Reply:
+1, this is why I went with the deepcopy, i built the first version without it for non-tuples and the side-effect on the original datapipe for dropping columns seemed incorrect given other patterns of use

Reply:
I think deep copy is needed if you choose to use `del new_item[i]`. What if we don't use `del` and construct a new object like this? (I think this is closer to what `MapperMapDataPipe` does):
```python
def __init__(self, source_datapipe: MapDataPipe, indices: list[int]):
    self.source_datapipe = source_datapipe
    self.indices = set(indices)

def __getitem__(self, index):
    old_item = self.source_datapipe[index]
    new_item = [x for i, x in enumerate(old_item) if i not in self.indices]
    if isinstance(old_item, tuple):
        return tuple(new_item)
    return new_item
```

Reply:
The only annoying thing about constructing instead of deleting is you need to do another split on dictionary vs. list in addition to the tuple check since you then have to also construct a dictionary differently, where deleting with an index key vs. a key from a dictionary is the same thing

Reply:
Take a look at the new version @NivekT and @ejguan, I've updated to build the items instead of using delete, it does end up with different list/tuple/dict options but looks ok to me. lmk what you think